code/game/objects/effects/effect_system.dm

```diff
@@ -392,17 +407,19 @@ steam.start() -- spawns the effect
 	if(holder)
 		src.location = get_turf(holder)
 	for(i=0, i<src.number, i++)
-		spawn()
-			var/obj/effect/effect/smoke/smoke = new smoke_type(location)
-			var/direction
-			if(cardinals)
-				direction = pick(cardinal)
-			else
-				direction = pick(alldirs)
+		var/obj/effect/effect/smoke/smoke = new smoke_type(location)
+		var/direction
+		if(cardinals)
+			direction = pick(cardinal)
+		else
+			direction = pick(alldirs)
+		var/added_time = 1 SECOND
+		for(i=0, i<pick(0,1,1,1,2,2,2,3), i++)
+			addtimer(CALLBACK(src, PROC_REF(move_smoke), smoke, direction), added_time)
```

Comment:
```suggestion
		for(var/j=0, j<pick(0,1,1,1,2,2,2,3), j++)
			addtimer(CALLBACK(src, PROC_REF(move_smoke), smoke, direction), added_time)
```
Spawn() allowed this potential infinite loop to work without breaking the server.